include/envoy/upstream/upstream.h

```diff
@@ -942,5 +959,43 @@ class Cluster {
 
 using ClusterSharedPtr = std::shared_ptr<Cluster>;
 
+/**
+ * A pool of upstream connections that all serving threads can offer or retrieve upstream
+ * connections from by Host.
+ */
+class UpstreamConnectionPool {
+public:
+  /**
+   * Sockets and additional information about the upstream connection being transferred.
+   */
+  struct UpstreamConnectionEssence {
+    Network::ConnectionSocketPtr socket;
+    Network::TransportSocketPtr transport_socket;
+    uint64_t remaining_requests;
+  };
+  virtual ~UpstreamConnectionPool() = default;
+  /**
+   * Transfer a connection from a serving thread's connection pool to the global shared connection
+   * pool.
+   * @param host supplies the host associated with the connection.
+   * @param connection_essence supplies the sockets associated with the upstream connection being
+   * transferred.
+   */
+  virtual void OfferConnection(HostConstSharedPtr host,
+                               UpstreamConnectionEssence connection_essence) PURE;
+  /**
+   * Transfer a connection from the global shared connection pool to a serving thread.
+   * @param host supplies the host associated with the desired connection.
+   * @param dispatcher supplies the dispatcher associated with the serving thread's event loop which
+   * will accept the connection.
+   * @param accept_cb supplies the callback to invoke in the serving thread to take ownership of the
+   * retried connection_essence, if any, or handle the retrieve failure. #return true if the
+   * asynchronous operation to retrieve a connection was started, or false if the retrieve operation
+   * is rejected because there are no available connections in the pool.
+   */
+  virtual bool RetrieveConnection(HostConstSharedPtr host, Event::Dispatcher& dispatcher,
+                                  std::function<void(UpstreamConnectionEssence)> accept_cb) PURE;
```

Comment:
Could be done by adding a thread-safe object that mediates the cancellation.  That said, does envoy cancel requests on cluster removal?  The current implementation could cause a crash on request cancellation, I'll write a test to cover that case and fix the APIs so that case is handled correctly.

Reply:
Yes in the case of a hard removal, all requests will be reset as the connection pool will be destroyed. See `ClusterManagerImpl::removeCluster`